src/components/course/routes/externalCourseEnrollmentLoader.js

```diff
@@ -0,0 +1,62 @@
+import { generatePath, redirect } from 'react-router-dom';
+
+import {
+  extractEnterpriseId,
+  getLateRedemptionBufferDays,
+  queryCanRedeem,
+  queryCourseMetadata,
+  queryRedeemablePolicies,
+} from '../../app/data';
+import { ensureAuthenticatedUser } from '../../app/routes/data';
+
+export default function makeExternalCourseEnrollmentLoader(queryClient) {
+  return async function externalCourseEnrollmentLoader({ params = {}, request }) {
+    const requestUrl = new URL(request.url);
+    const authenticatedUser = await ensureAuthenticatedUser(requestUrl, params);
+    // User is not authenticated, so we can't do anything in this loader.
+    if (!authenticatedUser) {
+      return null;
+    }
+
+    const {
+      enterpriseSlug,
+      courseType,
+      courseKey,
+      courseRunKey,
+    } = params;
+    const enterpriseId = await extractEnterpriseId({
+      queryClient,
+      authenticatedUser,
+      enterpriseSlug,
+    });
+
+    // Fetch course metadata, and then check if the user can redeem the course.
+    // TODO: This should be refactored such that `can-redeem` can be called independently
+    // of `course-metadata` to avoid an unnecessary request waterfall.
+    await queryClient.ensureQueryData(queryCourseMetadata(courseKey, courseRunKey)).then(async (courseMetadata) => {
+      const redeemableLearnerCreditPolicies = await queryClient.ensureQueryData(queryRedeemablePolicies({
+        enterpriseUuid: enterpriseId,
+        lmsUserId: authenticatedUser.userId,
+      }));
+      const isEnrollableBufferDays = getLateRedemptionBufferDays(redeemableLearnerCreditPolicies.redeemablePolicies);
+      const canRedeem = await queryClient.ensureQueryData(
+        queryCanRedeem(enterpriseId, courseMetadata, isEnrollableBufferDays),
+      );
+      const hasSuccessfulRedemption = !!canRedeem.find(r => r.contentKey === courseRunKey)?.hasSuccessfulRedemption;
```

Comment:
👍

Reply:
So, you'll notice the `hasSuccessfulRedemption` is now taking into account the `courseRunKey` route param (if present) in the custom hook that retrieves the metadata from the `can-redeem` API (i.e., `useCourseRedemptionEligibility`). That way, any consumer of the `useCourseRedemptionEligibility` can consume `hasSuccessfulRedemption` without needing to consider the optional `courseRunKey` route param. AKA the "same-named key in various dictionaries" will largely already have the `courseRunKey` filtering taken into account when accessing the property, so it's not really semantically different than the original `hasSuccessfulRedemption`.

That said, I opted to keep the data transform at the `useQuery` layer instead of at the query function (service) layer given the only transformed property needed at the route loader layer is `hasSuccessfulRedemption`; accepting a tiny bit of duplicate logic within the route loader as the `useQuery` data transform. Would be open to considering doing the data transforms that exist within `useCourseRedemptionEligibility` directly in the query function so it's applicable to both the route loader and `useQuery` by default so the `hasSuccessfulRedemption` doesn't need to be computed in 2 places, though.

Also, FWIW, we might want to also wire up that optional `?course_run_key` query parameter supported by the course about page route (i.e., where only the specific course run from the query param should be displayed). In that case, we'd likely also only want to use the specified `?course_run_key` within the `can-redeem` API query (rather than all available course run keys, since only 1 course run would be displayed anyways). The integration of the `?course_run_key` query parameter and the `can-redeem` API is not currently wired up as is.